### ISSUE TITLE
UnusedPrivateMember should ignore methods annotated with @PostConstruct / @PreDestroy

### ISSUE THREAD

**Issue by ilya40umov** [detekt/detekt#2231]
## Expected Behavior
UnusedPrivateMember should ignore methods annotated with @PostConstruct / @PreDestroy

## Current Behavior
Currently any method that is private and is annotated with @PostConstruct / @PreDestroy is still reported (even though it will be called via reflection at runtime by Spring anyway).

## Context
I get that I could rename all of the methods to be called like "postConstruct" / "preDestroy" and then add them to the excluded names, but it will harm the readability of the code (methods have a name so that it can define what the method does and not what it should be called).

**Comment by imanushin** [comment]
I think you are talking about Spring IoC and [about these annotations](https://www.baeldung.com/spring-postconstruct-predestroy).

If I'm right, I can recommend you replacing reflection-based approach with attributes to explicit interface implementation, e.g.:

* Instead of ```@PreDestroy ``` attribute you can use ```DisposableBean``` interface, e.g. you need to implement method ```destroy```
* Instead of ```@PostConstruct``` attribute you can use ```InitializingBean``` interface with ```afterPropertiesSet``` method.

After these changes Detekt warning will be disappeared.

If you insist on using attribute-based approach I advice extending configuration of this rule, to have ability of putting "ignored attributes" there. E.g. you will just able to put both attributes into this configuration, however Detekt will know nothing about these Spring attributes.

Looks like this issue can be related to #2178

**Comment by arturbosch** [comment]
I agree that we should provide annotation support for `UnusedPrivateMember`. Annotations like `@Test` of JUnit also extend the entry points of a program. The user should't need to change his/her code for such cases.

**Comment by ilya40umov** [comment]
@imanushin I would really prefer to avoid `DisposableBean` / `InitializingBean`, as they are essentially coming from the pre-annotation times and aren't really different from having a special name for the methods (that I would then have to mark as "allowed" in the detekt config).
I would say having an option to specify the list of annotations that indicate the methods that must be ignored by `UnusedPrivateMember` should much better.

**Comment by schalkms** [comment]
> I agree that we should provide annotation support for UnusedPrivateMember.

I disagree with your statement. @arturbosch 
I don’t think we should adjust this rule just to fit some Spring use cases. We didn‘t do this in the past. I still remember the code snippets written with the Dagger framework.

If we want to introduce this complexity (I am still not convinced), I can think of the following approach.
In my opinion this is more of a cross functional rule concern. I think it’s better to declare annotations globally. Every reported **KtElement** (no matter by which rule) should be checked against this annotation list.

**Comment by osipxd** [comment]
@schalkms, you mean to add global entry points annotations list?
Something like this?
```yml
config:
    entryPointAnnotations: "com.example.EventHandler"
```

**Comment by schalkms** [comment]
@osipxd exactly

**Comment by arturbosch** [comment]
@schalkms @osipxd I like your proposal.
We will also need such a property for a full deadcode analysis.

**Comment by BraisGabin** [comment]
After I read #2747 and reread this I have another proposal. Probably we need something similar of https://github.com/detekt/detekt/issues/2581#issuecomment-609496344:

```yml
exclusion-patterns:
- patterns:
    - **/test/**
    - **/androidTest/**
    - **/*.Test.kt
    - **/*.Spec.kt
    - **/*.Spek.kt
  rules:
    - VariableNaming
    - FunctionNaming
- patterns:
    - *DTO.kt
  rules:
    - Rule4
    - Rule8
...
```

In that issue we only talk about path pattern exclusion. But It could have sense to extend that proposal and add annotations too. What do you think?

**Comment by rock3r** [comment]
Just to add to this discussion, the unused members checks flag all Jetpack Compose-annotated `@Preview` functions, because they are technically unused (but the IDE tooling is using them). Would be great to be able to suppress (at least some) rules for members annotated with certain annotations.

**Comment by jordansilva** [comment]
Any update on this? What solution are you using for Jetpack Compose for example? @Ignore or making a Custom rule?

**Comment by jordansilva** [comment]
I think this PR will handle annotation problems.
https://github.com/detekt/detekt/pull/2734

**Comment by BraisGabin** [comment]
I think that the @rock3r's comment is the one with more upvotes this year so we should do somethig here (thank you a lot for all those who thumbed up it and don't commented with +1).

We know that the "right fix" will take a while to arrive. So I vote to fix this "the old way" and add a `ignoreAnnotated` to `UnusedPrivateMember` as `LongParameterList` or `LongMethod` have.

**Comment by dimsuz** [comment]
> Does anyone want to jump to it?

I've just done it and only then saw that @BraisGabin prepared a more generic approach to this problem.

I guess no point in making a PR now (it would add `ignoreAnnotated` only to the `UnusedPrivateMember`), but if I'm wrong and there's some reason to have this, ping me and I'll create one!
